radicle-node/src/service.rs

```diff
@@ -975,6 +985,48 @@ where
         Ok(false)
     }
 
+    /// A convenient method to check if we should fetch from a `RefsAnnouncement`
+    /// with `scope`.
+    fn should_fetch_refs_in_announcement(
+        &self,
+        message: &RefsAnnouncement,
+        scope: &tracking::Scope,
+    ) -> Result<bool, session::Error> {
+        let should_fetch = match scope {
+            tracking::Scope::All => true,
+            tracking::Scope::Trusted => {
+                match self.tracking.namespaces_for(&self.storage, &message.rid) {
+                    Ok(Namespaces::All) => true,
+                    Ok(Namespaces::Many(nodes)) => {
+                        // Get the set of trusted nodes except self.
+                        let my_id = self.node_id();
+                        let node_set: HashSet<_> =
+                            nodes.iter().filter(|key| *key != &my_id).collect();
+
+                        // Check if there is at least one trusted ref.
+                        message
+                            .refs
+                            .iter()
+                            .any(|(pub_key, _refs)| node_set.contains(pub_key))
+                    }
+                    Ok(Namespaces::One(key)) => {
+                        message.refs.iter().any(|(pub_key, _refs)| pub_key == &key)
+                    }
+                    Err(NamespacesError::NoTrusted { rid }) => {
+                        debug!(target: "service", "no trusted nodes to fetch {}", &rid);
+                        false
+                    }
+                    Err(e) => {
+                        error!(target: "service", "failed to obtain namespaces: {e}");
+                        return Err(e).map_err(session::Error::Namespaces);
+                    }
+                }
+            }
+        };
+
+        Ok(should_fetch)
+    }
```

Comment:
I'm not sure what you mean [here](https://github.com/radicle-dev/heartwood/pull/370/files#r1143965771) but this is what I'm talking about:
```suggestion
        match scope {
            tracking::Scope::All => Ok(true),
            tracking::Scope::Trusted => {
                match self.tracking.namespaces_for(&self.storage, &message.rid) {
                    Ok(Namespaces::All) => Ok(true),
                    Ok(Namespaces::Many(nodes)) => {
                        // Get the set of trusted nodes except self.
                        let my_id = self.node_id();
                        let node_set: HashSet<_> =
                            nodes.iter().filter(|key| *key != &my_id).collect();

                        // Check if there is at least one trusted ref.
                        Ok(message
                            .refs
                            .iter()
                            .any(|(pub_key, _refs)| node_set.contains(pub_key)))
                    }
                    Ok(Namespaces::One(key)) => {
                        Ok(message.refs.iter().any(|(pub_key, _refs)| pub_key == &key))
                    }
                    Err(NamespacesError::NoTrusted { rid }) => {
                        debug!(target: "service", "no trusted nodes to fetch {}", &rid);
                        Ok(false)
                    }
                    Err(e) => {
                        error!(target: "service", "failed to obtain namespaces: {e}");
                        Err(e.into())
                    }
                }
            }
        }
```

Reply:
This is simpler, I agree.

Reply:
Updated.